traits/trait_list_object.py

```diff
@@ -212,471 +156,290 @@ def notify(self, index, removed, added):
         for notifier in self.notifiers:
             notifier(self, index, removed, added)
```

Comment:
Should we catch errors raised by the user defined notifier so that the operation doesn't fail even if the notifier errors out ? There are cases where we return a value after calling `notify` and this would fail if any notification fails.

Reply:
This is a good question, though I think it's bigger than just this one PR. It's _possible_ that we want to catch all failed user notifiers and turn the failures into logged warnings. But that's dangerous: it risks masking coding errors, and we also have internal notifiers to worry about (in fact, that's most of the use-case right now).

So yes, right now we should get a loud failure (an exception and traceback) if any notification fails, and right now that's what we want, since a failed notification likely represents a coding error. (We may still want to be catching exceptions inside notifiers at the point where they call into user callbacks; I think @kitchoi 's work still does that.)

Reply:
Indeed, an imminent PR for EEP 3 is going to do the exception handling inside the notifiers.

In addition to the reasons Mark mentioned, this is consistent with how ctraits `call_notifiers` work (so the notifiers will have to do their error handling anyway). I did not appreciate this earlier but I do now: This allows fine grained control on error handling later in the notifiers. The notifiers do a pile of things before it calls the user-provided handler. We'd want to do a catch-all around the user's handler that is outside traits control, but not around code that belongs to traits.